File: GenerateParenthesis.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    const result = [];
    const createParenthesis = (str, strLength, open, close) => {
        if(strLength === n * 2) {
            result.push(str);
            return;
        }
        if(open < n) {
            createParenthesis(str + '(', strLength + 1, open + 1, close);
        }
        if(close < open) {
            createParenthesis(str + ')', strLength + 1, open, close + 1);

        }
    }
    createParenthesis('', 0, 0, 0);
    return result;
};

File: GenerateParenthesis.java
class Solution {
    
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        generateParenthesis(result, "", 0, 0, n);
        return result;
    }
    
    void generateParenthesis(List<String> result, String s, int open, int close, int max){
        if(s.length() == max * 2){
            result.add(s);
            return;
        }    
        
        if(open < max) generateParenthesis(result, s + "(", open + 1, close, max);
        if(close < open) generateParenthesis(result, s + ")", open, close + 1, max);
        
    }
    
}
